### PR TITLE
Closes #914 by handling case where the num_elements < num_locales

### LINKED ISSUE #914: Categorical/Strings array with size smaller than num locales fails read/write to HDF5
As identified in issue #904 we have an issue where both Strings & Categoricals appear to fail reading / writing to HDF5 when the size of the array is less than the number of locales.

This can be produced by modifying the unit test `tests/categorical_test.py::CategoricalTest::testSaveAndLoadCategorical` and setting the number of elements to something smaller than the number of locales.

Note: It is probably best to make an explicit unit test to cover this issue.
Reproducible via:
```python
# mkdir -p /tmp/junk
# ./arkouda_server -nl 2 --logLevel=LogLevel.DEBUG --GenSymIO.logLevel=LogLevel.DEBUG

import arkouda as ak
ak.connect("localhost")
c1 = ak.Categorical(ak.array(["abc"]))
c1.save("/tmp/junk/cat", dataset="c1_array")
x = ak.load_all("/tmp/junk/cat")
c2 = x["c1_array"]
```

NOTE: To enable more than one locale for local testing you need to turn on the following environment variables when compiling and running (see also chapel-lang docs for running multi-locale https://chapel-lang.org/docs/usingchapel/multilocale.html#readme-multilocale )
```
export CHPL_COMM=gasnet
export CHPL_COMM_SUBSTRATE=smp
export CHPL_RT_OVERSUBSCRIBED=yes

# It's a good idea to also turn on
export ARKOUDA_DEVELOPER=true
```

### PR DESCRIPTION
**Pull request by glitch**
Closes #914 by handling case where the num_elements < num_locales
when saving to HDF5. When data is not present for a locale an
empty dataset is written to the file. These empty datasets get ignored
when reading back from disk.

Added unit tests for small numeric & strings arrays where we
expect the number of locales to be larger then the number of elements.

### Notes
1. Most of the changes for big blocks of code are white-space related because they were wrapped in if/else conditionals to handle boundary conditions.  It may be easier to view these sections in an ide or visual diff tool that can ignore whitespace changes (I like pycharm for this personally)
2. The fundamental change here is adding empty datasets to files on write when data does not exist on a locale and ignoring those on read.  This allows dataset / groups to be present in all files so no errors get thrown while reading back and the `alllow_errors` and `strictTypes` flags do not need to be toggled by the user (potentially masking other errors).

### CODE REVIEW

**Comment by ronawho** [comment]
> may be easier to view these sections in an ide or visual diff tool that can ignore whitespace

FYI github supports this. In https://github.com/Bears-R-Us/arkouda/pull/934/files under the settings gear icon, or you can just do url hacking and add `?w=1`. e.g.  https://github.com/Bears-R-Us/arkouda/pull/934/files?w=1

File: src/GenSymIO.chpl [hunk 1]
```diff
@@ -650,10 +651,17 @@ module GenSymIO {
...
-            b = sd.low; // Boundary is index of first segment in file
-            // Height increase of next region is number of bytes in current region
-            if (i < fD.high) {
-                diffs[i+1] = vd.size;
+            // if we encounter a malformed subdomain i.e. {1..0} that means we encountered a file
+            // that has no data for this SegString object, we can safely skip processing this file.
+            if sd.low > sd.high {
+                gsLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                    "fD:%t segments subdom:%t is malformed signaling no segment data in file, skipping".format(i, sd));
+            } else {
```
Comment (hokiegeek2) [inline src/GenSymIO.chpl:659]: more chpldoc comments would be good in the else code block

File: src/GenSymIO.chpl [hunk 2]
```diff
@@ -930,6 +939,12 @@ module GenSymIO {
                 C_HDF5.HDF5_WAR.H5LTget_dataset_info_WAR(file_id, dName.c_str(), 
                                            c_ptrTo(dims), nil, nil);
                 lengths[i] = dims[0]: int;
+                if lengths[i] == 0 {
+                    skips.add(filename);
+                    try! gsLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
```
Comment (hokiegeek2) [inline src/GenSymIO.chpl:944]: recommend getting rid of try! as it can crash server

**Review by hokiegeek2 — APPROVED** [review]

**Review by reuster986 — CHANGES_REQUESTED** [review]
General logic looks good, just a couple minor requests.

**Review by reuster986 — APPROVED** [review]

File: src/GenSymIO.chpl [hunk 3]
```diff
@@ -646,14 +647,33 @@ module GenSymIO {
...
 
+    /**
+     * inline proc to validate the range for our domain.
+     * Valid domains must be increasing with the lower bound <= upper bound
+     * :arg r: 1D domain
+     * :type domain(1): one dimensional domain
+     *
+     * :returns: bool True iff the lower bound is less than or equal to upper bound
+     */
+    inline proc _isValidRange(r: domain(1)): bool {
```
Comment (ronawho) [inline src/GenSymIO.chpl:658]: Note that this is the same as `!r.isEmpty()` if you wanted to just use a builtin. In terms of terminology we don't tend to think of this as valid vs malformed, but just empty vs non-empty.

**Pull request merged**